app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt

```diff
@@ -32,4 +37,59 @@ sealed class Diagnostic {
                 append(" Expected symbols: ${expectedSymbols.joinToString()}.")
         }.toString()
     }
+
+    data class ConstantWithoutValue(val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "A constant must have a value"
+    }
+
+    data class UninitializedGlobalVariable(val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "A global variable must be initialized"
+    }
+
+    data class ImmutableAssignment(val assignment: Statement.Assignment, val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot assign to a " + if (variable.kind == Variable.Kind.CONSTANT) "constant" else "value"
+    }
+
+    data class ParameterAssignment(val assignment: Statement.Assignment, val parameter: Function.Parameter) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot assign to a parameter"
+    }
+
+    data class FunctionAssignment(val assignment: Statement.Assignment, val function: Function) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot assign to a function"
+    }
+
+    data class FunctionAsValue(val expression: Expression.Variable, val function: Function) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot use a function as a value"
+    }
+
+    data class VariableCall(val call: Expression.FunctionCall, val variable: Variable) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot call a variable"
+    }
+
+    data class ParameterCall(val call: Expression.FunctionCall, val parameter: Function.Parameter) : Diagnostic() {
+        override fun isError() = true
+        override fun toString() = "Cannot call a parameter"
+    }
```

Comment:
Do we need those diagnostics here? I think these errors are already checked in name resolution

Reply:
I think it should actually be the responsibility of the type checker, not the name resolution, to report these errors, because they are errors only for as long as functions are not first-class, and this is a property of the type system. In principle, variables, parameters and functions have the same namespace, and name resolution shouldn't have to care in what contexts these names are used.